### ISSUE TITLE
Infinite redraw when a variable is undefined.

### ISSUE THREAD

**Issue by v4n** [lhorie/mithril.js#1170]
### Description:

Experiencing an infinite redraw when a variable is undefined in the controller.

### Steps to Reproduce:

Created an example [codepen example](https://codepen.io/anon/pen/mELRrm?editors=0010).
Open the chrome console, uncomment the undefined variable in line 17 and then you should start seeing the console output an undefined error & the m.request firing forever.

I've been able only to reproduce this by using the layout function and having an undefined variable *after* the `m.request`.

**Comment by barneycarroll** [comment]
I've experienced this too but didn't make the connection with Mithril per se (thought it was a problem with async debugging)

**Comment by isiahmeadows** [comment]
@v4n @barneycarroll 

It's definitely a bug if you ever run into that problem, and it can be reduced pretty far ([see this pen for a configurable minimal repro](https://codepen.io/isiahmeadows/pen/oLdOxW?editors=0010)). I will note that it doesn't repro with `requestAnimationFrame` (and it likely never will in browsers with only `setTimeout` and `setInterval`).

But, here's why this isn't a Mithril bug: you scheduled the redraw to occur *after* the controller threw. Mithril doesn't know about that intent to redraw until after the controller threw. You *could* track thrown errors as well as returned results, but it doesn't make much sense to do so, since errors are much less common, and aren't really worth anything to rendering.

You'll likely get the same result with most vdom frameworks if you throw during controller instantiation after scheduling a function to request an asynchronous redraw (e.g. in React calling `this.replaceState` in a `setTimeout` callback within the constructor, but throwing at the end of it).

**Comment by v4n** [comment]
@isiahmeadows from the perspective of someone not familiar with mithril's internals it's a bug with the framework. 

On a language that is not type safe I wouldn't be expecting mithril to go into an infinite redraw if a undefined variable creeps after scheduling an async. redraw. On a page that has api requests and high amount visits this could potentially cause ddos to a server.

In my company we got automated testing so we would catch this before going live. That being said it's great when the framework protects the developer and has little amount of edge cases that one has to be in the look for.

**Comment by isiahmeadows** [comment]
@v4n The thing is Mithril doesn't have that much knowledge nor control of the browser event loop to mitigate this. This is a basic JS problem dealing with order of evaluation.

```js
// Don't run this. If you run this in Node, it'll hang the process immediately,
// logging `run` and `load` alternately, and you may have trouble killing it.
function load() {
  console.log("load")
  process.nextTick(run)
  throw new Error("Hello")
}

function run() {
  console.log("run")
  load()
}

run()
```

The reason this is unavoidable is because in order to prevent it, you have to intercept attempts to add to the event queue and *only* actually add them if the component returns successfully. Neither the browser, Node, nor any JS engine I'm aware of provides a non-trivial way to intercept any of these attempts at all, either in JS or native, beyond just setting a few internal callbacks to integrate them with an existing event loop before running anything. It's not even possible to get the internal callbacks after they've been set.

-------

Monkeypatching is a possible stopgap, but it can get difficult in a hurry with all the assumptions you have to catch, and it's not a proper fix. Consider [Bluebird's scheduler](https://github.com/petkaantonov/bluebird/blob/master/src/schedule.js) as an example of something that's relatively difficult to patch, and also consider the many native task scheduling methods in various browsers:

- `Promise.prototype.then` and `Promise.prototype.catch` in ES2015
- `setTimeout`
- `setInterval`
- `setImmediate`
- `process.nextTick` in Node.js
- Various setters and mutating methods on DOM nodes with an attached `MutationObserver`
- etc.

Some testing frameworks that allow a test to fail after a set timeout (e.g. Mocha) cache `setTimeout` to avoid Sinon's monkeypatching when you use their [fake timers API](http://sinonjs.org/docs/#clock) or [Lolex (which Sinon's uses under the hood)](https://github.com/sinonjs/lolex). And if native Promises exist, it's possible to use this to protect against patching:

```js
var p = Promise.resolve()
exports.schedule = p.then.bind(p)
```

-------

In summary, it's a JS safety limitation that's exceptionally difficult to work around, and in Mithril, I don't believe the need is worth the difficulty of implementing it. Also, we've already run into issues with detecting various methods early, such as Safari issues detailed in #917 deferring a `try-catch` to runtime in #923.  So I'm a major :-1:, and I suspect @lhorie and @tivac would likely concur here.
